controller_manager/src/controller_manager.cpp

```diff
@@ -1606,40 +1614,93 @@ controller_interface::return_type ControllerManager::update(
   auto ret = controller_interface::return_type::OK;
   ++update_loop_counter_;
   update_loop_counter_ %= update_rate_;
-
+  
+  
+  for (auto& wrapped_thread : async_controller_threads_) // map should be empty in the initial update call so this is fine
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    wrapped_thread.second.set_time_and_period(time, period);
+  }
+  
+  
   for (auto loaded_controller : rt_controller_list)
   {
     // TODO(v-lopez) we could cache this information
     // https://github.com/ros-controls/ros2_control/issues/153
     if (is_controller_active(*loaded_controller.c))
     {
       const auto controller_update_rate = loaded_controller.c->get_update_rate();
+      const auto controller_is_async = loaded_controller.c->is_async();
+
 
       bool controller_go =
         controller_update_rate == 0 || ((update_loop_counter_ % controller_update_rate) == 0);
-      RCLCPP_DEBUG(
-        get_logger(), "update_loop_counter: '%d ' controller_go: '%s ' controller_name: '%s '",
-        update_loop_counter_, controller_go ? "True" : "False",
+      RCLCPP_WARN(
+        get_logger(), "update_loop_counter: '%d ' is_async: '%s ' controller_name: '%s '",
+        update_loop_counter_, controller_is_async ? "True" : "False",
         loaded_controller.info.name.c_str());
 
       if (controller_go)
-      {
-        auto controller_ret = loaded_controller.c->update(
-          time, (controller_update_rate != update_rate_ && controller_update_rate != 0)
-                  ? rclcpp::Duration::from_seconds(1.0 / controller_update_rate)
-                  : period);
+      { 
+          if (controller_is_async) 
+          {
+            if (async_controller_threads_.find(loaded_controller.info.name) == async_controller_threads_.end()) // create new thread only if we didn't do it for an existing async controller
+            { //TODO: synchronize with controller_go
+              
+              async_controller_threads_.emplace(loaded_controller.info.name, ControllerThreadWrapper(time, period));
+              async_controller_threads_.at(loaded_controller.info.name).start([this, loaded_controller]()
+
+            
+              {
+                while (!terminated_)
```

Comment:
For maintainability I suggest to move this code to the controller wrapper class. I know it's cooler to do it with a lambda but we really don't need to present this logic here to the reader.